src/utils/logging.js

```diff
@@ -0,0 +1,83 @@
+/* 
+const silentLogger: ILogger = {
+    debug: noop,
+    info: noop,
+    warn: noop,
+    error: noop
+};
+
+let loggerInstance: ILogger = globalThis?.console;
+
+export function configureLogger(logger: null | ILogger) {
+    if (!logger) {
+        loggerInstance = silentLogger;
+        return;
+    }
+    loggerInstance = logger;
+};
+
+export const logger = {
+    debug(...data) {
+        loggerInstance.debug(...data);
+    },
+    info(...data) {
+        loggerInstance.info(...data);
+    },
+    warn(...data) {
+        loggerInstance.warn(...data);
+    },
+    error(...data) {
+        loggerInstance.error(...data);
+    }
+};
+*/
+
+function noop() { }
+
+const silentLogger = {
+    debug: noop,
+    info: noop,
+    warn: noop,
+    error: noop
+};
+
+let loggerInstance = globalThis?.console;
+
+/**
+ * Sets a logger object to be used by the single-spa library to emit log messages.  By default, 
+ * log messages are emitted to the console.
+ * @param logger A custom logger object, or `null` to silence logging.  To restore logging to 
+ * the console, pass the console object.
+ */
+export function configureLogger(logger) {
+    if (logger === null) {
+        loggerInstance = silentLogger;
+        return;
+    }
+    if (typeof logger === 'object' &&
+        typeof logger.debug === 'function'
+        && typeof logger.info === "function"
+        && typeof logger.warn === "function"
+        && typeof logger.error === "function"
+    ) {
+        loggerInstance = logger;
+    }
+    else {
+        throw new Error("Invalid argument:  The given logger does not conform to the expected specification.")
```

Comment:
I'd still prefer just logging a message to the console at the error level using `formatErrorMessage()`, so its inline with the other error messages. This is a programming error, but it shouldn't stop anyone's application.

Reply:
Hello.  I just searched for `formatErrorMessage()` and is meant to throw minified error messages.  Does this mean that you want a new error code to be created for this error?

Also, once the throw is removed, what do we do?  Not change the value of `loggerInstance` or do you want the value to revert to the console?  Or some other option?

Reply:
I'd think we'd just not change it at all, so, if for some reason, someone did:

```
configureLogger(customLogger); 

configureLogger(true);
```

They'd just get an error, but the custom logger would remain set. Likewise, just:

```
configureLogger(true);
```

With no other should just leave the normal logger (the console) attached.

> Does this mean that you want a new error code to be created for this error?

Yes.

Reply:
Good questions! The errors are defined [here](https://github.com/single-spa/single-spa.js.org/tree/master/src/components/Errors/Codes) (that's the website repo) and are mounted into the view [here](https://github.com/single-spa/single-spa.js.org/blob/master/src/pages/error/index.js) (so you can see the available properties, etc.) The next available error number is 42.

Ideally, the error message would be clear and as understandable as possible. E.g., it may make sense to pass in the actual argument that `configureLogger()` was called with (using `JSON.stringify()`), so on the app side we can display a clear, comprehensible message, e.g.,

```ts
configureLogger() was called with `true` which does not conform to the ILogger interface. An ILogger should be an object that has the following properties...
```

Reply:
> deally, the error message would be clear and as understandable as possible. E.g., it may make sense to pass in the actual argument that configureLogger() was called with (using JSON.stringify()), so on the app side we can display a clear, comprehensible message,

Hmmm, ok.  Understandable request but then I have some troublesome concerns with this:

1. Data privacy.  This could potentially expose a client's private/corporate data to single-spa.js.org.  Actually, worse because the data is in the query string, so it could be sniffed without any hassles while in transit.
2. It could serve as a DoS attack?  I'm thinking that if the passed object is large, stringifying it could be used to attack the single-spa website.  I'm not a security guy, so please correct me if you know better.
3. If we decided we still wanted to stringify (personally I'd prefer not to), should the result be trimmed to a maximum length?  I think it should.

Other than that, I guess I'll make a PR for the website project.